### DISCUSSION TITLE
Specify how background tasks are handled on behalf of the function being async or not

### DISCUSSION THREAD

**Discussion by espetro** [fastapi/fastapi#8875] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed 

### Example

```python
import time
import asyncio

from fastapi import FastAPI, BackgroundTasks

app = FastAPI()

def sync_bg_task(name: str) -> str:
    time.sleep(5)
    return f"Hello {name}"

async def async_bg_task(name: str) -> str:
    await asyncio.sleep(5)
    return f"Hello {name}"

@app.get("/{name}")
async def hello_name_sync(name: str, background_tasks: BackgroundTasks):
    background_tasks.add_task(sync_bg_task, name)
    return "OK"

@app.get("/async/{name}")
async def hello_name_async(name: str, background_tasks: BackgroundTasks):
    background_tasks.add_task(async_bg_task, name)
    return "OK"
```

### Description

The internal behaviour of background tasks changes on behalf of the function being defined as async or not. [In the docs][docs], it says that _it can be an async def or normal def function, FastAPI will know how to handle it correctly_. However, that is disguising some information.
As explained [in this other issue][bug], an async function background task will run in asyncio loop (thus, everything that doesn't use "await" will block); however, a normal function background will run in a threadpool.

**Can this be explicitly written [in the docs][docs]? I do not consider it as a bug but as a very important detail the developer must know.**

### Environment

* OS: macOS
* FastAPI Version: 0.65.1
* Python version: 3.9.1

### Additional context

There is a [related thread where this is pointed as a bug][bug].

[bug]: https://github.com/tiangolo/fastapi/issues/1313#issuecomment-618206969
[docs]: https://fastapi.tiangolo.com/tutorial/background-tasks/?h=backgr#create-a-task-function

**Comment by raphaelauv** [comment]
ok , open a Pull Request

**Comment by CA-Lee** [comment]
I'd follow someone's path and had found a hacking way: add `asyncio.sleep(0.001)` as a backgroung task before others async functions, which eventually make it behalf like sync background tasks. Perhaps you may consider to add this in the docs. :)

ref. https://github.com/tiangolo/fastapi/issues/1313#issuecomment-751355711

**Comment by iamthen0ise** [comment, marked as answer]
Afaik BackgroundTask is just a proxy to Starlette' one. 
So the behaviour is pretty straightforward, and existing explanation is pretty descriptive.

(https://github.com/encode/starlette/blob/master/starlette/background.py#L6-L20)

```python3
class BackgroundTask:
    def __init__(
        self, func: typing.Callable, *args: typing.Any, **kwargs: typing.Any
    ) -> None:
        self.func = func
        self.args = args
        self.kwargs = kwargs
        self.is_async = asyncio.iscoroutinefunction(func)


    async def __call__(self) -> None:
        if self.is_async:
            await self.func(*self.args, **self.kwargs)
        else:
            await run_in_threadpool(self.func, *self.args, **self.kwargs)
```
